New tutorial: how to protect a REST endpoint (#91)
kytos/documentation: tutorials/napps/how_to_protect_a_napp_rest_endpoint.rst

:tocdepth: 2
:orphan:

.. _tutorial-how-to-protect-a-napp-rest-endpoint:

###################################
How to create your own NApp: Part 4
###################################

********
Overview
********

In this tutorial, you will learn how to create a Network Application (NApp) for
*Kytos* (|kytos|_) to communicate with your application. This NApp will be made
with a public and a private REST endpoints.

The average time to go through this is: ``20 min``.

What you will learn
===================

* How to create a superuser;
* How to create a NApp with a public and a private REST endpoint.

What you will need
==================

* How to create a basic NApp: Part 1 - Refer to |Tutorial_01|_;
* How to create a basic NApp: Part 2 - Refer to |Tutorial_02|_;
* How to create a basic NApp: Part 3 - Refer to |Tutorial_03|_;
* Read the documentation: |protect_rest|_;
* Your |dev_env|_ already up and running;
* The NApp Storehouse, installed and running.

************
Introduction
************

> Beyond the events, one of the methods to communicate with your applications are

Reviewer: ```suggestion
Besides events, another method to communicate with your NApp is
```

> through REST endpoints. These endpoints can be public and private, this last is

Reviewer: ```suggestion
through REST endpoints. These endpoints can be public or private (note that
```

> still under development.

Reviewer: ```suggestion
the latter is still under development).
```

********

In this tutorial, you will learn how to create a Network Application (NApp) for
*Kytos* (|kytos|_) to communicate with your application. This NApp will be made
with a public and a private REST endpoints.

The average time to go through this is: ``20 min``.

What you will learn
===================

* How to create a superuser;
* How to create a NApp with a public and a private REST endpoint.

What you will need
==================

* How to create a basic NApp: Part 1 - Refer to |Tutorial_01|_;
* How to create a basic NApp: Part 2 - Refer to |Tutorial_02|_;
* How to create a basic NApp: Part 3 - Refer to |Tutorial_03|_;
* Read the documentation: |protect_rest|_;
* Your |dev_env|_ already up and running;
* The Storehouse NApp, installed and running.

************
Introduction
************

Besides events, another method to communicate with your NApp is
through REST endpoints. These endpoints can be public or private
(please note that the latter is still under development).

The creation of a new endpoint is made by the use of the `@rest` decorator,
and it can be optionally protected using the `@authenticated`
decorator. More information about this module can be found in

> the |protect_rest|_ and in the Auth documentation.

Reviewer: Is there another page with the Auth documentation whose link is missing from here? Please add link.

your application:

.. code-block:: python

    from kytos.core import KytosNApp, log, rest
    from napps.<username>.hellorest import settings
    from flask import jsonify, request

    class Main(KytosNApp):

        def setup(self):
            self.name = "REST NApp"

        def execute(self):
            log.info("Running!")

        def shutdown(self):
            log.info("Bye!")
        
        @rest('v1/name', methods=['GET'])
        def get_name(self):
            return jsonify(self.name), 200

Run kytos:

.. code-block:: console

    $ kytosd -f

Open another terminal and install your application:

.. code-block:: console

    $ python setup.py develop

Check if your application is installed and running with the command:

.. code-block:: console

    $ kytos napps list

    Status |          NApp ID          |              Description
    =======+===========================+=======================================
    [ie]  | kytos/storehouse          | ….
    [ie]  | kytos/hellorest           | ….

Accessing the REST endpoint of your application in a browser or REST client,
you will receive a message with a return like this:

.. code-block:: console

>     GET http://0.0.0.0:8181/api/<username>/<napp_name>/name

Reviewer: Please add examples directly as curl commands on the shell.

    {"REST NApp"}

How to protect your endpoint
============================

Now, you can secure the created endpoint. To do this, you have to
import the `@authenticated` decorator and add it to this endpoint method.

.. code-block:: python

    from kytos.core import KytosNApp, log, rest, authenticated
    from napps.<username>.hellorest import settings
    from flask import jsonify, request

    class Main(KytosNApp):

        def setup(self):
            self.name = "REST NApp"

        def execute(self):
            log.info("Running!")

        def shutdown(self):
            log.info("Bye!")
        
        @rest('v1/name', methods=['GET'])
        @authenticated
        def get_name(self):
            return jsonify(self.name), 200

Then, if you try to access this endpoint, you will receive an error message
requesting the authentication token. This token must be obtained in the
superuser login process.

.. code-block:: console

>     GET http://0.0.0.0:8181/api/<username>/<napp_name>/name

Reviewer: curl

Now that superuser has been created, you should access the login endpoint to
request the access token:

.. code-block:: console

    GET http://0.0.0.0:8181/api/kytos/core/auth/login/

    $ curl -X GET \
        -H 'Accept:application/json' \
        -H 'Authorization:Basic <username>:<password>' \
        URL

You will receive the token:

.. code-block:: console

    {
        "token": <token>
    }

Now, you have to send the token in the `Bearer` HTTP header
to be able to access this endpoint:

.. code-block:: console

    GET http://0.0.0.0:8181/api/<username>/<napp_name>/name

    $ curl -X GET \
        -H 'Content-type: application/json' \
        -H 'Accept: application/json' \
        -H 'Authorization: Bearer ${TOKEN}' \
        -d '{"user_id": "001"}' \
        URL

If no error occurs, you will receive the normal return from the endpoint:

.. code-block:: console

    {"REST NApp"}

Note: If an error occurs, check the error message and request a new token if
it has expired.

Congratulations! You have completed the tutorial and explored the creation of REST
endpoints using different request methods like POST, PATCH and DELETE. One tip is

> that you can use a REST client named Postman to test your application.

Reviewer: Please add a link to the Postman app page.
